text/0000-startup-initialized-statics.md

```diff
@@ -0,0 +1,97 @@
+
+### 2. Function attribute: `startup_initialization`
+```rust
+#[startup_initialization]
+fn init_my_data() {
+    DATA[0] = 11;
+}
+```
+The `startup_initialization` attribute takes no arguments and it can only be attached to a function (a so-called *"startup initialization function"*) that takes no arguments and has an empty return. A user cannot call such a function directly, but a call can happen implicitly and only either in the beginning of main function or after one or more calls to other startup initialization functions, the first of which being called in the beginning of main. All startup initialized statics, which have a given startup initialization function as the argument to their `startup_initialized` attribute (the so called *"associated startup initialized statics"*), are considered mutable inside the body of said function so that they can be initialized by using runtime functionality. A startup initialization function may also access (either directly or indirectly through a function call) other (non-associated) startup initialized statics, even if they're in other crates, but the dependency on other startup initialized statics must not be circular. For example, given the startup initialization functions `func_x` and `func_y`, and their respective associated startup initialized statics `data_x` and `data_y`, if `func_x` accesses `data_y` and `func_y` accesses `data_x`, then this forms a circular dependency which would result in a compile-time error indicating the responsible startup initialization functions. All function calls made by a startup initialization function (either directly or indirectly) must be statically dispatched. The previous restriction is not strictly necessary, and could be relaxed later to allow dispatching dynamically given that the compiler is able to exhaustively list all the functions that could possibly be called. Marking a startup initialization function as `pub` should result in a warning saying something like *"visibility has no bearing on startup initialization functions"*.
+
+## A possible implementation
+
+It's getting tedious to write *"startup initialization"* and *"startup initialized"*, so from now on, I'll shorten both of them to SI.
+
+Inside the body of an SI function, all of its associated SI statics are made mutable. It is as if the following kind of transformation happened:
+
+From:
+```rust
+type Data = [i8; 3];
+
+#[startup_initialized(init_my_data)]
+static DATA: Data = [1,2,3];
+
+#[startup_initialization]
+fn init_my_data() {
```

Comment:
@kennytm This wouldn't get rid of the unsafe transformation (i.e. transmute to &mut) but only hoist it to the caller. And like the RFC said, inside the body of a startup initialization function, **all** of its associated startup initialized statics are made mutable. So, you might need more than one argument. Also, it would obfuscate what exactly is being mutated:
```rust
type Data = [i8; 3];

#[startup_initialized(init_my_data)]
static DATA_1: Data = [1,2,3];

#[startup_initialized(init_my_data)]
static DATA_2: Data = [4,5,6];

#[startup_initialization]
fn init_my_data(x: &mut Data, y: &mut Data) {
    x[0] = 11; // what does x refer to?
    y[0] = 22; // what does y refer to?
}
```